cg/server/admin.py

```diff
@@ -429,6 +430,57 @@ def view_sample_link(unused1, unused2, model, unused3):
             else ""
         )
 
+    @action(
+        "cancel_samples",
+        "Cancel samples",
+        "Are you sure you want to cancel the selected samples?",
+    )
+    def cancel_samples(self, entry_ids: List[str]) -> None:
+        all_associated_case_ids = set()
+
+        for entry_id in entry_ids:
+            sample = db.sample_by_id(id_=entry_id)
+
+            sample_case_ids = [case_sample.family.internal_id for case_sample in sample.links]
+            all_associated_case_ids.update(sample_case_ids)
+
+            db.delete_relationships_sample(sample=sample)
+            self.write_cancel_comment(sample=sample)
+
+        case_ids = list(all_associated_case_ids)
+        db.delete_cases_without_samples(case_ids=case_ids)
+        cases_with_remaining_samples = db.get_cases_with_samples(case_ids=case_ids)
+
+        self.display_cancel_confirmation(entry_ids, cases_with_remaining_samples)
+
+    def write_cancel_comment(self, sample: Sample) -> None:
+        username = db.user(session.get("user_email")).name
+        date = datetime.now().strftime("%Y-%m-%d")
+        comment = f"Cancelled {date} by {username}"
+
+        db.set_sample_comment(sample=sample, comment=comment)
+
+    def display_cancel_confirmation(self, entry_ids: List[str], remaining_cases: List[str]) -> None:
+        samples = "sample" if len(entry_ids) == 1 else "samples"
+        cases = "case" if len(remaining_cases) == 1 else "cases"
+
+        message = f"Cancelled {len(entry_ids)} {samples}. "
+        case_message = ""
+
+        for case_id in remaining_cases:
+            case_message = case_message + f" {case_id},"
+
+        case_message = case_message.strip(",")
```

Comment:
Since it will be added one extra time, there is probably a better way. Otherwise the message will be ```case_id_1, case_id_2,```. The last one should be removed.